ibl_to_nwb/brainwide_map/read_after_write.py

```diff
@@ -0,0 +1,361 @@
+import os
+import numpy as np
+import pandas as pd
+from pathlib import Path
+import h5py
+from pynwb import NWBHDF5IO, NWBFile
+from one.api import ONE
+import logging
+from typing import Optional
+from numpy import testing
+
+def check_arrays(array_a: np.ndarray, array_b: np.ndarray, full_check: bool = False):
```

Comment:
This style of redirection is not recommended (a helper function just to decide between uses of `numpy.testing`)

TBH any data array in the processed files will be small enough such that a full (non-subset) usage of `numpy.testing.assert_allclose` should always be fast enough to just do that

The only question for submitting would be if you want to do something like this on the raw data, but that's probably important enough for it to be a separate check altogether

Also, the NWB write step should not affect floating point precision (and if it did we would actually want to be informed about that) so I recommend using `numpy.testing.assert_array_equal` instead. Or in general, recommend `numpy.testing.assert_array_almost_equal` since it should check other aspects of the arrays beyond the data values

Reply:
> The only question for submitting would be if you want to do something like this on the raw data, but that's probably important enough for it to be a separate check altogether

I had this in mind when I was writing the function, and that small speed improvements are probably going to result in large total time savings. That said, I haven't done any benchmarking yet (besides the `np.searchsorted` way to access the spike times).

> Also, the NWB write step should not affect floating point precision (and if it did we would actually want to be informed about that) so I recommend using numpy.testing.assert_array_equal instead. Or in general, recommend numpy.testing.assert_array_almost_equal since it should check other aspects of the arrays beyond the data values

I wasn't entirely sure if this is true across all architectures or if we need to consider some other form of rounding error imprecision, so I thought using `allclose` would allow to adjust for that if necessary, and if true equality is desired `rtol=0` should do that.

Reply:
One way to think of the NWB conversion process (for bulk data anyway) is that 'bytes in' are still just 'bytes out', so if you decoded the original SpikeGLX binaries into their raw bytes (things that look like `\xff\x01r`), those sequences should be exactly the same as when you read them from the data array after it's written to the NWB file

Floating point errors would mostly come up if you performed operations like scaling (such as by the conversion factor to get values into scientific units)

Or if you used lossless compression to save more space in the file; though that would often exceed standard 'floating point' error